src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,527 @@
+  }
+
+  /**
+   * Returns true if the user is interacting with the carousel.
+   */
+  isBeingInteractedWith(): boolean {
+    return this.transition.isInteracting();
+  }
+
+  /**
+   * Returns the currently active slide.
+   */
+  getActiveSlide(): HTMLElement {
+    return this.transition.getActiveSlide();
+  }
+
+  /**
+   * Return the index of the currently active slide.
+   */
+  getActiveSlideIndex(): number {
+    return this.getSlideIndex(this.getActiveSlide());
+  }
+
+  /**
+   * Return the first slide in the carousel.
+   */
+  getFirstSlide(): HTMLElement {
+    return this.slides[0];
+  }
+
+  /**
+   * Return the last slide in the carousel.
+   */
+  getLastSlide(): HTMLElement {
+    return this.slides[this.slides.length - 1];
+  }
+
+  /**
+   * Return the index of the given slide within the list of slides.
+   * @param slide
+   */
+  getSlideIndex(slide: HTMLElement): number {
+    return this.getSlides().indexOf(slide);
+  }
+
+  /**
+   * Return the carousel container, the DOM element containing the slides.
+   */
+  getContainer(): HTMLElement {
+    return this.container;
+  }
+
+  /**
+   * Return a copy of the list of slide elements.
+   */
+  getSlides(): HTMLElement[] {
+    return [...this.slides];
+  }
+
+  /**
+   * Return the slides before the given slide. Looping if allowed.
+   * @param slide
+   */
+  getSlidesBefore(slide: HTMLElement): HTMLElement[] {
+    if (this.allowsLooping()) {
+      return this.splitSlidesInHalf(slide, Direction.LEFT);
+    } else {
+      return this.getSlides().slice(0, this.getSlides().indexOf(slide));
+    }
+  }
+
+  /**
+   * Return the slides after the given slide. Looping if allowed.
+   * @param slide
+   */
+  getSlidesAfter(slide: HTMLElement): HTMLElement[] {
+    if (this.allowsLooping()) {
+      return this.splitSlidesInHalf(slide, Direction.RIGHT);
+    } else {
+      return this.getSlides().slice(this.getSlides().indexOf(slide) + 1);
+    }
+  }
+
+  /**
+   * Transition to the next slide.
+   */
+  next(): void {
+    this.transitionSlidesBy(1);
+  }
+
+  /**
+   * Transition to the previous slide.
+   */
+  previous(): void {
+    this.transitionSlidesBy(-1);
+  }
+
+  /**
+   * Transition to the slide `value` slides away.
+   * @param value
+   */
+  transitionSlidesBy(value: number): void {
+    const nextIndex =
+        this.getSlides().indexOf(this.getCurrentTransitionTarget()) + value;
+    this.transitionToIndex(nextIndex);
+  }
+
+  /**
+   * Transition to the slide with the given index.
+   *
+   * @param index
+   * @param drivenBySync
+   */
+  transitionToIndex(index: number, drivenBySync: boolean = false): void {
+    const clampedIndex = this.getClampedIndex(index);
+    this.transitionToSlide(this.getSlideByIndex(clampedIndex), drivenBySync);
+  }
+
+  /**
+   * Register a function to be called when the carousel starts a transition.
+   * @param callback
+   */
+  onTransitionStart(callback: (carousel: Carousel) => void) {
```

Comment:
No establish pattern since there aren't many UI components.

lazy-image and a few others do fire events on the root dom element itself.  Example: https://github.com/blinkkcode/degu/blob/main/src/dom/lazy-video.ts#L205

Reply:
Setup EventManager class and EventDispatcher interface per our chat.